compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -1729,7 +1780,17 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             .filter(|r| self.trait_defines_associated_const_named(r.def_id(), assoc_name));
 
         let (bound, next_cand) = match (matching_candidates.next(), const_candidates.next()) {
-            (Some(bound), _) => (bound, matching_candidates.next()),
+            (Some(bound), _) => {
+                let next = matching_candidates.next();
+                // Automatically prefer `T: Trait` over `T: ~const Trait` for back compat
+                if let Some(bound) =
+                    next.and_then(|bound2| bound.pick_concrete_effect(bound2, constness))
```

Comment:
why do we actually have multiple candidates instead of a single `for<E> T host<E> Trait`?

Reply:
uh 🤦 🤯 🤦

Reply:
or i guess idk if that works :shrug:

Reply:
Ah no, it can't work, we need it to be host-on for non-const functions. Don't want to require const impls in non const mode